### PR TITLE
[CLS-2209] Server TLS check

### PR DESCRIPTION
**Pull request by KtorZ**
Note that new certificates have been generated using the [scripts provided on daedalus](https://github.com/input-output-hk/daedalus/tree/master/installers) in a slightly modified configuration. I'll open a PR on daedalus once this one is accepted :+1:

### CODE REVIEW

**Comment by adinapoli-iohk** [comment]
@KtorZ Mind pleasing the CI gods on this one? After that, I will review, I promise 😛

**Comment by KtorZ** [comment]
Haha yeah, I was honestly waiting for your review first, such that if there's any changes, I don't have to make an offering to the CI gods twice :3

**Comment by adinapoli-iohk** [comment]
@KtorZ Also a very good point! Will get to it then 😉

File: lib/src/Pos/Web/Server.hs [hunk 1]
```diff
@@ -98,10 +106,60 @@ serveImpl app host port mWalletTLSParams mSettings =
...
+    mTlsConfig = tlsParamsToWai host port <$> mWalletTLSParams
 
-tlsParamsToWai :: TlsParams -> TLSSettings
-tlsParamsToWai TlsParams{..} = tlsSettingsChain tpCertPath [tpCaPath] tpKeyPath
+tlsParamsToWai
+    :: String -> Word16 -> TlsParams -> TLSSettings
+tlsParamsToWai host port TlsParams{..} = tlsSettingsWithCertCheck
+  where
+    tlsSettings =
+      tlsSettingsChain tpCertPath [tpCaPath] tpKeyPath
```
Comment (disassembler) [inline lib/src/Pos/Web/Server.hs]: I should have reviewed this one first, this answers my question about intermediates :)

File: wallet/node/NodeOptions.hs [hunk 2]
```diff
@@ -114,4 +115,26 @@ tlsParamsOption = do
...
+          "tlsclients"
+          "LIST"
+          "A comma-separated list of accepted clients"
+          <> Opt.value ["Daedalus Wallet"]
     return TlsParams{..}
+
+
+strList :: Opt.ReadM [String]
+strList =
+    Opt.str >>= failIfEmpty . splitOn ','
```
[the remark this answers is not in the document]
Comment (KtorZ) [reply wallet/node/NodeOptions.hs]: It is. I wasn't sure about what was best... redefining this small function quickly to work with strings, or convert to text, use Text.splitOn, convert back to string.
I guess you stand on the second approach ?

[the remark this answers is not in the document]
Comment (adinapoli-iohk) [reply wallet/node/NodeOptions.hs]: I think that for non-performance-critical code like this one, the double conversion is justified by the increase in DRYness 😁

**Comment by vincenthz** [comment]
I may have missed something fundamental, but I don't understand how this is suppose to work; Is it that the server is started with the right client allowed list in command line when daedalus generate a new client server ?

Also just to be sure, what's the certificate commited in the repo for ?

**Comment by KtorZ** [comment]
@vincenthz The idea is to start the server with a given CA certificate that is used to sign both the server certificate and the one of Daedalus (or any clients that would connect to the server). 

To authenticate a client, the server would first verify that the client certificate is actually valid and was signed by a common authority, and then check the name of the client identified by the certificate.

The name check has no special use for now, but we might have different types of clients later, having access to a different APIs. It is pretty much useless for now (and I am actually considering removing it, won't be much to re-introduce later when needed...).

---

The certificates present & versionned on the repo are simply dev certificates. The clients one can be used with *cURL* or *Postman* to interact with the server. The installers / launchers for Daedalus or Cardano-node would have to generate fresh new certificates and ensure that the generate private keys are only readable by the appropriate applications. This is ongoing work with the OPS and will follow #2742.

Inputs are welcome.

**Comment by KtorZ** [comment]
@adinapoli-iohk I am sorry, I've completely passed by your comments from last week. To be honest, since this client name verification isn't really useful for now, I am also tempted of simply removing it for now. 

If I understand you correctly, you are thinking of activating the TLS server-side check at runtime, depending on the CLI that are provided? Could do indeed. 

About the certificates, they fix an issue with the Subject Alternative Names that are present on the current certificates. This isn't mandatory for the check to work as expected thus, I don't see any reason why we couldn't also revert back for now to the previous certs. 
Yet, since they're only dev certificates, I wonder if it matters much :/ ?

**Pull request closed without merging**